### PR TITLE
Addition of interface code

### PR DESCRIPTION
**Pull request by stromjm**
Master branch as of July 2024 was modified to enable the interface extraction mode that was first created in the interface branch in December 2023. This original branch was not able to be successfully merged with master, so this is a fresh start. Additional change from the original branch is the new parameter exposing the distance threshold for defining the interface.

### CODE REVIEW

File: src/strucclustutils/structcreatedb.cpp [hunk 1]
```diff
@@ -15,6 +15,7 @@
 #include "PatternCompiler.h"
 #include "Coordinate16.h"
 #include "itoa.h"
+#include <tmalign/basic_fun.h>
```
Comment (milot-mirdita) [inline src/strucclustutils/structcreatedb.cpp:18]: Please remove this and add a dist function to `MathUtil`

File: src/strucclustutils/structcreatedb.cpp [hunk 2]
```diff
@@ -113,12 +114,13 @@ void addMissingAtomsInStructure(GemmiWrapper &readStructure, PulchraWrapper &pul
...
 void findInterfaceResidues(GemmiWrapper &readStructure, std::pair<size_t, size_t> res1, std::pair<size_t, size_t> res2,
-                           std::vector<size_t> & resIdx1, std::vector<size_t> & resIdx2)
+                           std::vector<size_t> & resIdx1)
 {
     std::vector<Vec3> coord1, coord2;
     LocalParameters& par = LocalParameters::getLocalInstance();
     for (size_t res1Idx = res1.first; res1Idx < res1.second; res1Idx++) {
-        float shortestDistance = std::numeric_limits<float>::infinity();
+        // float shortestDistance = std::numeric_limits<float>::infinity();
+        float squarethreshold = par.distanceThreshold * par.distanceThreshold;
```
Comment (milot-mirdita) [inline src/strucclustutils/structcreatedb.cpp:123]: Please pull this above the for-loop and mark it const

Comment (milot-mirdita) [reply src/strucclustutils/structcreatedb.cpp:123]: Also it should be squareThreshold

File: src/strucclustutils/structcreatedb.cpp [hunk 3]
```diff
@@ -142,79 +144,107 @@ void findInterfaceResidues(GemmiWrapper &readStructure, std::pair<size_t, size_t
...
-            D2 += (z1 - z2) * (z1 - z2);
-            float distance = sqrt(D2);
-            if (distance < shortestDistance) {
-                shortestDistance = distance;
+            // float D2 = 0;
+            // D2 += (x1 - x2) * (x1 - x2);
+            // D2 += (y1 - y2) * (y1 - y2);
+            // D2 += (z1 - z2) * (z1 - z2);
+            // float distance = sqrt(D2);
+            float distance = BasicFunction::dist(x1, y1, z1, x2, y2, z2);
```
Comment (milot-mirdita) [inline src/strucclustutils/structcreatedb.cpp:152]: Please move to `MathUtil` and don't include a tmalign header

File: src/strucclustutils/structcreatedb.cpp [hunk 4]
```diff
@@ -142,79 +144,107 @@ void findInterfaceResidues(GemmiWrapper &readStructure, std::pair<size_t, size_t
...
-void compute3DiInterfaces(GemmiWrapper &readStructure, StructureTo3Di &structureTo3Di, SubstitutionMatrix & mat3Di) {
+void compute3DiInterfaces(GemmiWrapper &readStructure, StructureTo3Di &structureTo3Di, SubstitutionMatrix & mat3Di, int chainNameMode) {
     std::vector<size_t> resIdx1, resIdx2;
+    std::vector<int> interfacetaxIds;
+    std::vector<unsigned int> interfaceModelIndices;
     std::vector<Vec3> ca, n, c, cb;
     std::vector<char> interfaceSeq3di, interfaceAmi;
     std::vector<Vec3> interfaceCa;
-    std::vector<std::string> interfaceNames, interfaceChainName;
+    std::vector<std::string> interfaceNames, interfaceChainNameVec;
```
Comment (milot-mirdita) [inline src/strucclustutils/structcreatedb.cpp:168]: I would suggest `interfaceChainNames`

File: src/strucclustutils/structcreatedb.cpp [hunk 5]
```diff
@@ -239,15 +273,14 @@ writeStructureEntry(SubstitutionMatrix & mat, GemmiWrapper & readStructure, Stru
...
-                    std::map<std::string, std::pair<size_t, unsigned int>> & entrynameToFileId,
+                    size_t & fileidCnt, std::map<std::string, std::pair<size_t, unsigned int>> & entrynameToFileId,
                     std::map<std::string, size_t> & filenameToFileId,
                     std::map<size_t, std::string> & fileIdToName,
                     DBWriter* mappingWriter) {
     LocalParameters &par = LocalParameters::getLocalInstance();
     if (par.dbExtractionMode == LocalParameters::DB_EXTRACT_MODE_INTERFACE) {
         addMissingAtomsInStructure(readStructure, pulchra);
-        compute3DiInterfaces(readStructure, structureTo3Di, mat);
+        compute3DiInterfaces(readStructure, structureTo3Di, mat, chainNameMode );
```
Comment (milot-mirdita) [inline src/strucclustutils/structcreatedb.cpp:283]: Please remove the extra space

**Review by milot-mirdita — COMMENTED** [review]
Minor style comments

File: src/strucclustutils/structcreatedb.cpp (hunk 2 above)
Comment (sooyoung-cha) [reply src/strucclustutils/structcreatedb.cpp:123]: yep

File: src/strucclustutils/structcreatedb.cpp (hunk 3 above)
Comment (sooyoung-cha) [reply src/strucclustutils/structcreatedb.cpp:152]: yep!

File: src/strucclustutils/structcreatedb.cpp (hunk 4 above)
Comment (sooyoung-cha) [reply src/strucclustutils/structcreatedb.cpp:168]: yep

File: lib/mmseqs/src/commons/MathUtil.h [hunk 6]
```diff
@@ -256,6 +256,13 @@ class MathUtil {
         return sum;
     }
 
+    static float squareDist(const float xx, const float xy, const float xz,
+               const float yx, const float yy, const float yz){
+        float d1=xx-yx;
```
Comment (milot-mirdita) [inline lib/mmseqs/src/commons/MathUtil.h:261]: code style: please add spaces:
```
float d1 = xx - yx;

File: src/commons/LocalParameters.cpp [hunk 7]
```diff
@@ -9,6 +9,8 @@ const int LocalParameters::DBTYPE_TMSCORE = 102;
 LocalParameters::LocalParameters() :
         Parameters(),
         PARAM_PREF_MODE(PARAM_PREF_MODE_ID,"--prefilter-mode", "Prefilter mode", "prefilter mode: 0: kmer/ungapped 1: ungapped, 2: nofilter",typeid(int), (void *) &prefMode, "^[0-2]{1}$"),
+        PARAM_DB_EXTRACTION_MODE(PARAM_DB_EXTRACTION_MODE_ID,"--db-extraction-mode", "Createdb extraction mode", "createdb extraction mode: 0: chain 1: interface",typeid(int), (void *) &dbExtractionMode, "^[0-1]{1}$"),
```
Comment (milot-mirdita) [inline src/commons/LocalParameters.cpp:12]: style: please add spaces after `,`:
```
PARAM_DB_EXTRACTION_MODE_ID, "--db-extraction-mode",
...
", typeid(int)
```

File: src/strucclustutils/structcreatedb.cpp [hunk 8]
```diff
@@ -89,117 +91,331 @@ std::string removeModel(const std::string& input) {
...
+                                    chainLen);
+        }
+    }
+}
+
+void findInterfaceResidues(GemmiWrapper &readStructure, std::pair<size_t, size_t> res1, std::pair<size_t, size_t> res2,
+                           std::vector<size_t> & resIdx1)
+{
+    std::vector<Vec3> coord1, coord2;
+    LocalParameters& par = LocalParameters::getLocalInstance();
```
Comment (milot-mirdita) [inline src/strucclustutils/structcreatedb.cpp:121]: Please pass in the parameters as function parameters, don't request the parameter singleton so deep in a function

File: src/strucclustutils/structcreatedb.cpp [hunk 9]
```diff
@@ -89,117 +91,331 @@ std::string removeModel(const std::string& input) {
...
+    size_t prevInterfaceChainLen = 0;
+    std::vector<char> interfaceSeq3di, interfaceAmi;
+    std::vector<size_t> resIdx1, resIdx2;
+    std::vector<int> interfacetaxIds;
+    std::vector<unsigned int> interfaceModelIndices;
+    std::vector<Vec3> ca, n, c, cb;
+    std::vector<Vec3> interfaceCa;
+    std::vector<std::string> interfaceNames, interfaceChainNames;
+    std::vector<std::pair<size_t, size_t>> interfaceChain;
+    std::map<unsigned int, unsigned int> modelToInterfaceNum;
```
Comment (milot-mirdita) [inline src/strucclustutils/structcreatedb.cpp:166]: Can this be an `unordered_map`?

File: src/strucclustutils/structcreatedb.cpp [hunk 10]
```diff
@@ -89,117 +91,331 @@ std::string removeModel(const std::string& input) {
...
+                    prevInterfaceChainLen += resIdx2.size();
+                    std::string interfaceName;
+                    if (Util::endsWith(".gz", readStructure.names[ch1] )){
+                        interfaceName.append(Util::remove_extension(Util::remove_extension(readStructure.names[ch1])));
+                    } else{
+                        interfaceName.append(Util::remove_extension(readStructure.names[ch1]));
+                    } 
+                    if(readStructure.modelCount > 1){
+                        interfaceName.append("_MODEL_");
+                        interfaceName.append(std::to_string(readStructure.modelIndices[ch1]));
```
Comment (milot-mirdita) [inline src/strucclustutils/structcreatedb.cpp:214]: Please use our SSTR, intead of `std::to_string`

File: src/strucclustutils/structcreatedb.cpp [hunk 11]
```diff
@@ -89,117 +91,331 @@ std::string removeModel(const std::string& input) {
...
+                    prevInterfaceChainLen += resIdx1.size();
+                    interfaceChain.push_back(std::make_pair(prevInterfaceChainLen, prevInterfaceChainLen + resIdx2.size()));
+                    prevInterfaceChainLen += resIdx2.size();
+                    std::string interfaceName;
+                    if (Util::endsWith(".gz", readStructure.names[ch1] )){
+                        interfaceName.append(Util::remove_extension(Util::remove_extension(readStructure.names[ch1])));
+                    } else{
+                        interfaceName.append(Util::remove_extension(readStructure.names[ch1]));
+                    } 
+                    if(readStructure.modelCount > 1){
```
Comment (milot-mirdita) [inline src/strucclustutils/structcreatedb.cpp:212]: style: whitespace

File: lib/mmseqs/src/commons/MathUtil.h (hunk 6 above)
Comment (sooyoung-cha) [reply lib/mmseqs/src/commons/MathUtil.h:261]: yep

File: src/commons/LocalParameters.cpp (hunk 7 above)
Comment (sooyoung-cha) [reply src/commons/LocalParameters.cpp:12]: yep

File: src/strucclustutils/structcreatedb.cpp (hunk 8 above)
Comment (sooyoung-cha) [reply src/strucclustutils/structcreatedb.cpp:121]: yep

File: src/strucclustutils/structcreatedb.cpp (hunk 9 above)
Comment (sooyoung-cha) [reply src/strucclustutils/structcreatedb.cpp:166]: sure

**Comment by milot-mirdita** [comment]
You can do an interactive rebase (git rebase -i) and squash the three commits into one and the do git push --force

**Comment by stromjm** [comment]
> @stromjm can you please undo the last two commits with a force push? Otherwise we might end up with keeping a lot of unrelated changes in the git history forever

yep, sorry about that, fixed

**Pull request merged**